File: leetcode/树/从前序与中序遍历序列构造二叉树.js
var buildTree = function (preorder, inorder) {
  const dfs = (arr1, arr2) => {
    if (!arr1.length) return null;
    let root = new TreeNode(arr1[0]);
    let mid = arr2.indexOf(arr1[0]);
    root.left = dfs(arr1.slice(1, mid + 1), arr2.slice(0, mid + 1));
    root.right = dfs(arr1.slice(mid + 1), arr2.slice(mid + 1));
    return root;
  };
  return dfs(preorder, inorder);
};

File: leetcode/树/从前序与中序遍历序列构造二叉树.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        dict={}
        for i in range(len(preorder)):
            dict[inorder[i]]=i
        def build(preorder,prestart,preend,inorder,instart,inend,dict):
            if prestart>preend or instart>inend:
                return None
            root=TreeNode(preorder[prestart])
            inroot=dict[preorder[prestart]]
            numLeft=inroot-instart
            root.left=build(preorder,prestart+1,prestart+numLeft,inorder,instart,inroot-1,dict)
            root.right=build(preorder,prestart+numLeft+1,preend,inorder,inroot+1,inend,dict)
            return root
        ans=build(preorder,0,len(preorder)-1,inorder,0,len(inorder)-1,dict)
        return ans